cmd/sink-worker/main.go

```diff
@@ -163,42 +181,54 @@ func main() {
 		logger.Error("failed to initialize sink worker", "error", err)
 		os.Exit(1)
 	}
-
-	var group run.Group
+	defer sink.Close()
 
 	// Set up telemetry server
-	{
-		server := &http.Server{
-			Addr:    conf.Telemetry.Address,
-			Handler: telemetryRouter,
-		}
-		defer server.Close()
-
-		group.Add(
-			func() error { return server.ListenAndServe() },
-			func(err error) { _ = server.Shutdown(ctx) },
-		)
+	server := &http.Server{
+		Addr:    conf.Telemetry.Address,
+		Handler: telemetryRouter,
+		BaseContext: func(_ net.Listener) context.Context {
+			return ctx
+		},
 	}
+	defer server.Close()
 
-	// Starting sink worker
-	{
-		defer sink.Close()
+	var group run.Group
+	// Add sink worker to run group
+	group.Add(
+		func() error { return sink.Run(ctx) },
+		func(err error) { sink.Close() },
+	)
 
-		group.Add(
-			func() error { return sink.Run() },
-			func(err error) { _ = sink.Close() },
-		)
-	}
+	// Add telemetry server to run group
+	group.Add(
+		func() error { return server.ListenAndServe() },
+		func(err error) { _ = server.Shutdown(ctx) },
+	)
 
 	// Setup signal handler
 	group.Add(run.SignalHandler(ctx, syscall.SIGINT, syscall.SIGTERM))
 
-	err = group.Run()
-	if e := (run.SignalError{}); errors.As(err, &e) {
-		slog.Info("received signal; shutting down", slog.String("signal", e.Signal.String()))
-	} else if !errors.Is(err, http.ErrServerClosed) {
-		logger.Error("application stopped due to error", slog.String("error", err.Error()))
+	// Run actors
+	if err = group.Run(); err != nil {
+		signalErr := &run.SignalError{}
+		if errors.As(err, signalErr) {
+			logger.Warn("received signal: shutting down", slog.String("signal", signalErr.Signal.String()))
+			switch signalErr.Signal {
+			case syscall.SIGTERM:
+				os.Exit(0)
+			default:
+				os.Exit(130)
+			}
+		}
+
+		if !errors.Is(err, http.ErrServerClosed) {
+			logger.Error("application stopped due to error", slog.String("error", err.Error()))
+			os.Exit(1)
+		}
 	}
```

Comment:
What I meant is keep the original structure. 😄 

`if e := (run.SignalError{}); errors.As(err, &e) {`